File: javascript/242.valid-anagram.js
/*
 * @lc app=leetcode id=242 lang=javascript
 *
 * [242] Valid Anagram
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function (s, t) {
  if (s.length !== t.length) return false;

  // Solution 1: inputs only consist of lowercase English letters
  const chars = new Array(26).fill(0);
  const d = "a".charCodeAt();
  for (let i = 0; i < s.length; i++) {
    chars[s.charCodeAt(i) - d]++;
    chars[t.charCodeAt(i) - d]--;
  }
  return chars.findIndex((c) => c !== 0) === -1;

  // Solution 2: inputs contain Unicode characters
  // const chars = new Map();
  // for (let i = 0; i < s.length; i++) {
  //   const u = chars.get(s[i]) || 0;
  //   chars.set(s[i], u + 1);
  //   const v = chars.get(t[i]) || 0;
  //   chars.set(t[i], v - 1);
  // }
  // for (let t of chars.values()) {
  //   if (t !== 0) return false;
  // }
  // return true;
};
// @lc code=end

File: rust/242.valid-anagram.rs
/*
 * @lc app=leetcode id=242 lang=rust
 *
 * [242] Valid Anagram
 */

// @lc code=start
impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        if s.len() != t.len() {
            return false;
        }

        let mut chars = vec![0; 26];
        for (a, b) in s.bytes().zip(t.bytes()) {
            chars[(a - b'a') as usize] += 1;
            chars[(b - b'a') as usize] -= 1;
        }
        return !chars.iter().any(|x| *x != 0);
    }
}
// @lc code=end
